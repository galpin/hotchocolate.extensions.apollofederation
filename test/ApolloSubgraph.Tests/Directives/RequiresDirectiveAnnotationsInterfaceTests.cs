using System.Threading.Tasks;
using HotChocolate.Types;
using Microsoft.Extensions.DependencyInjection;
using Xunit;
using static HotChocolate.Extensions.ApolloSubgraph.Test;

namespace HotChocolate.Extensions.ApolloSubgraph.Directives;

public class RequiresDirectiveAnnotationsInterfaceTests
{
    [Fact]
    public async Task When_provides_is_specified_on_interface()
    {
        var schema = await BuildSchemaAsync(builder =>
        {
            builder.AddType<IReview>();
            builder.AddType<Review>();
            builder.AddType<Product>();
            builder.AddQueryType();
        });

        var sut = schema.GetType<InterfaceType>(nameof(IReview));

        Assert.Collection(
            sut.Fields["product"].Directives,
            x => AssertEx.Directive(x, "requires", ("fields", "\"upc\"")));
        await schema.QuerySdlAndMatchSnapshotAsync();
    }

    public interface IReview
    {
        [GraphQLKey]
        int Id { get; }

        [GraphQLRequires("upc")]
        Product Product { get; }
    }

    public class Review : IReview
    {
        public Review(int id, Product product)
        {
            Id = id;
            Product = product;
        }

        [GraphQLKey]
        public int Id { get; }

        [GraphQLRequires("upc")]
        public Product Product { get; }
    }

    public class Product
    {
        public Product(string upc)
        {
            Upc = upc;
        }

        public string Upc { get; }
    }
}