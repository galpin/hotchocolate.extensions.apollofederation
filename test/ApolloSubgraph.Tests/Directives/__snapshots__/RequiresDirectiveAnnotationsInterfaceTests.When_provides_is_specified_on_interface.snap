{
  "data": {
    "_service": {
      "sdl": "interface IReview @key(fields: \u0022id\u0022) {\n  id: Int!\n  product: Product! @requires(fields: \u0022upc\u0022)\n}\n\ntype Product {\n  upc: String!\n}\n\ntype Review implements IReview @key(fields: \u0022id\u0022) {\n  id: Int!\n  product: Product! @requires(fields: \u0022upc\u0022)\n}"
    }
  }
}
